massa-versioning-worker/src/versioning_factory.rs

```diff
@@ -0,0 +1,405 @@
+use massa_time::MassaTime;
+use std::collections::BTreeMap;
+use std::iter;
+use thiserror::Error;
+
+use crate::versioning::{MipComponent, MipState, MipStateTypeId, MipStore};
+
+/// Factory error
+#[allow(missing_docs)]
+#[derive(Error, Debug)]
+pub enum FactoryError {
+    #[error("Unknown version, cannot build obj with version: {0}")]
+    UnknownVersion(u32),
+    #[error("Version: {0} is in versioning store but not yet implemented")]
+    UnimplementedVersion(u32), // programmer error, forgot to implement new version in factory traits
+    #[error("Version: {0} is not active yet")]
+    OnStateNotReady(u32),
+    #[error("Could not create object of type {0}: {1}")]
+    OnCreate(String, String),
+}
+
+#[derive(Clone)]
+/// Strategy to use when creating a new object from a factory
+pub enum FactoryStrategy {
+    /// use get_best_version (see Factory trait)
+    Best,
+    /// Require to create an object with this specific version
+    Exact(u32),
+    /// Create an object given a timestamp (e.g slot)
+    At(MassaTime),
```

Comment:
What's the usecase of this ?

Reply:
One use case is when you can to create a new block but in the future. If the state for component version 2 is locked in, in the future it can be active and you will want to use this version (2)